packages/xo-server/src/index.js

```diff
@@ -140,18 +147,44 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  let currentUser
   const SIGNIN_STRATEGY_RE = /^\/signin\/([^/]+)(\/callback)?(:?\?.*)?$/
   express.use(async (req, res, next) => {
-    const { url } = req
+    const { url, body } = req
     const matches = url.match(SIGNIN_STRATEGY_RE)
 
     if (matches) {
       return passport.authenticate(matches[1], async (err, user, info) => {
+        if (user) {
+          currentUser = user
+        }
+
         if (err) {
           return next(err)
         }
 
-        if (!user) {
+        // OTP authent
+        if (
+          currentUser !== undefined &&
+          currentUser.preferences !== undefined &&
+          currentUser.preferences.otp !== undefined
```

Comment:
You can use [optional chaining](https://babeljs.io/docs/en/next/babel-plugin-proposal-optional-chaining) in `xo-server`:

```js
const otp = currentUser?.preferences?.otp
```